src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "wallet/coinselection.h"
+#include "util.h"
+#include "utilmoneystr.h"
+
+// Descending order comparator
+struct {
+    bool operator()(const CInputCoin& a, const CInputCoin& b) const
+    {
+        return a.txout.nValue > b.txout.nValue;
+    }
+} descending;
+
+/*
+ * This is the Branch and Bound Coin Selection algorithm designed by Mark Erhardt. It is an exact match algorithm where the exact match
+ * is a range with the lower bound being the value we want to spend and the upper bound being that value plus the additional cost
+ * required to create and spend a change output. To do this, the algorithm builds a binary tree where each node is a UTXO and whether
+ * that UTXO is included or not in the current coin selection. The tree is searched in include-first order. Each UTXO is first included
+ * and the current selection is evaluated for whether it is within the target range. If it is over the threshold, we try excluding
+ * the UTXO previously added and the branch of the tree involving its inclusion is then not explored. This process is repeated until
+ * a solution is found (i.e. selected value falls within the range), the exhaustion limit is reached, or the tree is exhausted and no
+ * solution was found.
+ *
+ * To find the best possible solution, we use a waste metric. The waste metric is defined as the cost to spend the current inputs
+ * now minus the cost to spend the current inputs later, plus the amount exceeding the target value. We search the tree to find the
+ * set of UTXOs which falls within our range and minimizes waste.
+ *
+ * An additional optimization of this algorithm implemented here is a lookahead value which maintains the total value of the UTXO set
+ * of all unexplored UTXOs (i.e. UTXOs that have not yet been included or excluded). This allows us to cut a branch if the remaining
+ * amount is not sufficient to reach our target.
```

Comment:
I'd like to suggest the following for the above description:


This is the Branch and Bound Coin Selection algorithm designed by Murch. It searches for an
input set that can pay for the spending target and does not exceed this by more than the cost
of creating and spending a change output. The algorithm uses a depth-first search on a binary tree.
In the binary tree, each node corresponds to inclusion or omission of a UTXO. UTXOs are sorted by 
their effective values and the trees is explored per the inclusion branch first. At each node,
the algorithm checks whether the selection is within the target range. While the selection has
not reached the target range, more UTXOs are included. When a selection's value exceeds the target
range, the complete subtree deriving from this selection can be omitted. At that point, the last 
included UTXO is deselected and the corresponding omission branch explored instead. 
The search ends after the complete tree has been searched or when a limited number of tries has 
been tried.

The search does continue after the first solution has been found. The best solution is selected 
by minimizing the waste metric.  The waste metric is defined as the cost to spend the current 
inputs at the given fee rate minus the long term expected cost to spend the inputs, plus the amount 
the selection exceeds the spending target:

   waste = selectionTotal - target + inputs × (currentFeeRate - longTermFeeRate)

The algorithm contains two additional optimizations. A lookahand keeps track of the total value of 
the unexlored UTXOs. A subtree is not explored if the lookahead indicates that the target range
cannot be reached. Further, it is unnecessary to test equivalent combinations. This allows to skip 
testing the inclusion of UTXOs that match the effective value and waste as an omitted predecessor.

Reply:
s/unexlored/unexplored/

s/Further, it is unnecessary to test equivalent combinations./Further, it is unnecessary to test equivalent effective value combinations./

s/This allows to skip/This allows us to skip/

ACK otherwise

Reply:
Using this